Jint.Benchmark/UncacheableExpressionsBenchmark.cs

```diff
@@ -0,0 +1,192 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Text;
+using BenchmarkDotNet.Attributes;
+using BenchmarkDotNet.Configs;
+using BenchmarkDotNet.Diagnosers;
+using BenchmarkDotNet.Jobs;
+using Jint.Native;
+using Newtonsoft.Json;
+using Undefined = Jint.Native.Undefined;
+
+namespace Jint.Benchmark
+{
+    /// <summary>
+    /// Test case for situation where object is projected via filter and map, Jint deems code as uncacheable.
+    /// </summary>
+    [Config(typeof(Config))]
+    public class UncacheableExpressionsBenchmark
```

Comment:
Looks like a special case benchmark. If there is a bug (like something should be cached and is not) we should have a unit test. If this is just to make a specific case fast enough, then it shouldn't be in the repos, or at least on a custom branch where the work for improvement will happen, then be removed once it's done.

Reply:
In a way a special case, but this should reflect function invocation with arrays in quite generic way. Arrays struggle with these kinds of data where index access/traversal goes via dictionary instead of pure array indexing and produces more arrays.

There isn't a bug per se, but a performance problem when you try to filter/project large nested arrays. But feel free to remove if you want.

Reply:
`map` and `filter` are _very_ common things to do on arrays.
And they can usually be lifted quite easily, that is not just a generic scenario, that is something that I think should be optimized specifically

Reply:
Then these benchmarks should be in the same table as the sunspider ones, unless there are already included there, or within one dedicated to Array operations.

Reply:
In an Array benchmark class that would look like this http://jsben.ch/k3QoV. It doesn't have to be in this PR. We can do it later or as part of a perf PR that would focus on arrays.